esphome/components/tuya/tuya.cpp

```diff
@@ -105,39 +106,53 @@ void Tuya::handle_char_(uint8_t c) {
 }
 
 void Tuya::handle_command_(uint8_t command, uint8_t version, const uint8_t *buffer, size_t len) {
-  uint8_t c;
+  uint8_t c[] = {0x00, 0x00};
   switch ((TuyaCommandType) command) {
     case TuyaCommandType::HEARTBEAT:
       ESP_LOGV(TAG, "MCU Heartbeat (0x%02X)", buffer[0]);
       if (buffer[0] == 0) {
         ESP_LOGI(TAG, "MCU restarted");
-        this->send_empty_command_(TuyaCommandType::QUERY_STATE);
+        this->init_state_ = TuyaInitState::INIT_HEARTBEAT;
+      }
+      if (this->init_state_ == TuyaInitState::INIT_HEARTBEAT) {
+        this->init_state_ = TuyaInitState::INIT_PRODUCT;
+        this->send_empty_command_(TuyaCommandType::PRODUCT_QUERY);
       }
       break;
-    case TuyaCommandType::QUERY_PRODUCT: {
-      // check it is a valid string
-      bool valid = false;
+    case TuyaCommandType::PRODUCT_QUERY: {
+      // check it is a valid string made up of printable characters
+      bool valid = true;
       for (int i = 0; i < len; i++) {
-        if (buffer[i] == 0x00) {
-          valid = true;
+        if (! isprint(buffer[i])) {
```

Comment:
```suggestion
        if (!std::isprint(buffer[i])) {
```

Whitespace doesn't make travis happy - and std:: just for good measure.